app/src/org/commcare/views/widgets/CalendarFragment.java

```diff
@@ -46,45 +47,44 @@
 
     private long todaysDateInMillis;
     private static final int DAYSINWEEK = 7;
-    private static final String TIME = "TIME";
 
     @Override
-    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState){
+    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
         getDialog().getWindow().requestFeature(Window.FEATURE_NO_TITLE);
         inflateView(inflater, container);
 
-        if(savedInstanceState != null && savedInstanceState.containsKey(TIME)){
-            calendar = Calendar.getInstance();
-            calendar.setTimeInMillis(savedInstanceState.getLong(TIME));
-        }
-
         disableScreenRotation();
         initDisplay();
         initWeekDays();
         initOnClick();
-        refresh();
         setWindowSize();
 
         return layout;
     }
 
-    private void initWeekDays(){
+    @Override
+    public void onViewStateRestored(@Nullable Bundle savedInstanceState) {
+        super.onViewStateRestored(savedInstanceState);
+        refresh();

```

Comment:
Fix number 3: Don't call refresh() until after the prior state of the fragment has settled, or else that process overrides the new values